src/runners/junitRunner/JUnitRunnerResultAnalyzer.ts

```diff
@@ -117,10 +117,13 @@ export class JUnitRunnerResultAnalyzer extends RunnerResultAnalyzer {
             if (!this.currentItem) {
                 return;
             }
-
-            const testMessage: TestMessage = new TestMessage(this.traces);
-            this.tryAppendMessage(this.currentItem, testMessage, this.currentTestState);
             this.recordingType = RecordingType.None;
+            if (this.assertionFailure) {
+                this.tryAppendMessage(this.currentItem, this.assertionFailure, this.currentTestState);
+            }
+            if (this.traces?.value) {
+                this.tryAppendMessage(this.currentItem, new TestMessage(this.traces), this.currentTestState);
+            }
```

Comment:
2: add a potential assertionFailure and the stack-trace message just once, after we are done processing the stack-trace (and thus determined the best location to put it)